### PR TITLE
feat(sweeps): optuna scheduler job for launch sweeps

### PR DESCRIPTION
**Pull request by gtarpenning**
optuna scheduler job. 

Invoke a very basic Optuna example using the command: 

```bash
wandb launch-sweep optuna_scheduler/example_configs/basic.yaml --queue <queue> --project <project> --entity <entity>
```

Create your own optuna scheduler job by running: `python optuna_scheduler/optuna_scheduler.py` (with flags `--enable_git` for a git job, or when run in a container it should create an image job) 

This results in a nicely formatted output log that returns a job path, to be used in a launch sweep config: 

Screen Shot 2023-05-31 at 10 01 50 AM
Screen Shot 2023-05-31 at 10 05 12 AM

### CODE REVIEW

File: jobs/sweep_schedulers/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+   name: validation_accuracy
+   goal: maximize
+
+# Use a training job here, NOT what we just created above
+# This is the job to be used in the execution of the actual sweep
+job: <TRAINING JOB>
+
+# This is where we use the scheduler job that is created in the run above
+scheduler:
+   job: <SCHEDULER JOB>
```
Comment (KyleGoyette) [inline jobs/sweep_schedulers/README.md:39]: Should we prepopulate with the publicly available version

File: jobs/sweep_schedulers/README.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+   `wandb launch-sweep <config.yaml> --queue <queue> --project <project>`
+
+### Important Notes: 
+
+1. There are **two** different jobs that must be included in the sweep config! One is the training job, which can be created by running a local wandb run that has a call to `run.log_code()` (or is run inside of a container with the `WANDB_DOCKER` environment variable set). The second job is the one created by running the schedulers in this folder (job creation automatically handled). 
+2. For the `wandb_scheduler.py`, set the `method` of the sweep (bayes, grid, random) in the `scheduler.settings.method` key. All sweep schedulers sourced from jobs require `method: custom` in the top-level of the sweep configuration.
+
+### Optuna 
+
+More information specific to the Wandb-Optuna sweep scheduler can be found in the `wandb/examples` repo [here](https://github.com/wandb/examples/launch/launch-sweeps/)
```
Comment (KyleGoyette) [inline jobs/sweep_schedulers/README.md:64]: I think we link to the examples near the top, users coming here should be given access to a super fast way to run a sweep on launch. Then be introduced to the ways they can configure it.
Thoughts?

**Review by KyleGoyette — COMMENTED** [review]
Overall, I felt like I had to jump around alot between examples, and the jobs when running through for the first time. We should provide a single place (examples?) to introduce to running a basic sweep on launch. Then we should dive into more details.

I also think it would be good to have a README specific to the optuna scheduler that will explain more about it.

File: jobs/sweep_schedulers/README.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+```bash
+wandb launch-sweep optuna_scheduler/example_configs/basic.yaml --queue <queue> --project <project> --entity <entity>
+```
+
+More information specific to the Optuna sweep scheduler can be found in the `wandb/examples` repo [here](https://github.com/wandb/examples/tree/master/examples/launch/launch-sweeps/optuna-scheduler)
+
+### Custom scheduler jobs
+
+While the above example uses a pre-made scheduler job (`'wandb/jobs/Wandb Scheduler Image Job:latest'`), they can also be created using this repo. Clone the repo, modifying the `*_scheduler` scripts with impunity, and then create jobs from them, with:
```
Comment (KyleGoyette) [inline jobs/sweep_schedulers/README.md:46]: Nice, very simple start, now talk about customization

File: jobs/sweep_schedulers/README.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+`python wandb_scheduler/wandb_scheduler.py` or `python optuna_scheduler/optuna_scheduler.py`
+
+For a specific example of a simple modification to the `wandb_scheduler.py` file, check out the example in `wandb/examples` [here](https://github.com/wandb/examples/tree/master/examples/launch/launch-sweeps/custom-scheduler).
+
+Note: There are three possible job-types that can be created from this script, in the following ways:
```
Comment (KyleGoyette) [inline jobs/sweep_schedulers/README.md:52]: Wording is weird here, maybe:
Note: there are 3 possible job source types that one can use to create a job:

**Review by KyleGoyette — APPROVED** [review]

**Comment by gtarpenning** [comment]
I can't figure out how to rebase from remote and keep local changes... It says `fatal: Cannot rebase onto multiple branches.` when rebasing, where when merging it will just happily merge and squash commits.... ffs the git history is so scuffed now.

**Comment by contramundum53** [comment]
We tried optuna scheduler, but got the following error:
```
$ wandb launch-agent -e <USER_NAME> -q <QUEUE_NAME>
...
wandb: launch: Launching job: griffin_wb/public/job-wandb_job_sweep_scheduler_optuna:latest
wandb: launch: agent 8l2txup5 running 1 out of a maximum of 1 jobs
wandb:   2 of 2 files downloaded.
wandb: ERROR launch: Error running job: Traceback (most recent call last):
wandb: ERROR   File "/Users/c-bata/sandbox/wandb-launcher/venv/lib/python3.10/site-packages/wandb/sdk/launch/utils.py", line 653, in pull_docker_image
wandb: ERROR     docker.run(["docker", "pull", docker_image])
wandb: ERROR   File "/Users/c-bata/sandbox/wandb-launcher/venv/lib/python3.10/site-packages/wandb/docker/__init__.py", line 150, in run
wandb: ERROR     raise DockerError(
wandb: ERROR wandb.docker.DockerError: The docker command executed was `docker pull wandb/job_sweep_scheduler_optuna`.
wandb: ERROR It returned with code 1
wandb: ERROR The content of stdout is 'Using default tag: latest
wandb: ERROR '
wandb: ERROR The content of stderr is 'Error response from daemon: pull access denied for wandb/job_sweep_scheduler_optuna, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
wandb: ERROR '
wandb: ERROR
wandb: ERROR
wandb: ERROR During handling of the above exception, another exception occurred:
wandb: ERROR
wandb: ERROR Traceback (most recent call last):
wandb: ERROR   File "/Users/c-bata/sandbox/wandb-launcher/venv/lib/python3.10/site-packages/wandb/sdk/launch/agent/agent.py", line 461, in thread_run_job
wandb: ERROR     self._thread_run_job(launch_spec, job, default_config, api, thread_id)
wandb: ERROR   File "/Users/c-bata/sandbox/wandb-launcher/venv/lib/python3.10/site-packages/wandb/sdk/launch/agent/agent.py", line 508, in _thread_run_job
wandb: ERROR     run = backend.run(project, builder)
wandb: ERROR   File "/Users/c-bata/sandbox/wandb-launcher/venv/lib/python3.10/site-packages/wandb/sdk/launch/runner/local_container.py", line 114, in run
wandb: ERROR     pull_docker_image(image_uri)
wandb: ERROR   File "/Users/c-bata/sandbox/wandb-launcher/venv/lib/python3.10/site-packages/wandb/sdk/launch/utils.py", line 655, in pull_docker_image
wandb: ERROR     raise LaunchError(f"Docker server returned error: {e}")
wandb: ERROR wandb.sdk.launch.errors.LaunchError: Docker server returned error: The docker command executed was `docker pull wandb/job_sweep_scheduler_optuna`.
wandb: ERROR It returned with code 1
wandb: ERROR The content of stdout is 'Using default tag: latest
wandb: ERROR '
wandb: ERROR The content of stderr is 'Error response from daemon: pull access denied for wandb/job_sweep_scheduler_optuna, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
wandb: ERROR '
wandb: ERROR
wandb: ERROR
wandb: launch: agent 8l2txup5 polling on queues try-optuna-launcher-queue, running 0 out of a maximum of 1 jobs
```

Reproduction steps:
1. `pip install git+https://github.com/wandb/wandb.git`
2. `pip install optuna`
3. `wandb login`
4. In wandb dashboard, create a new project and a new queue
5. `git clone https://github.com/wandb/launch-jobs.git`
6. `cd launch-jobs/jobs/sweep-scheduler`
7. `wandb launch-agent -e <USER_NAME> -q <QUEUE_NAME>`
8. `wandb launch-sweep optuna_scheduler/example_configs/basic.yaml --queue <QUEUE_NAME> --project <PROJECT_NAME> --entity <USER_NAME>`

Maybe we are testing it in a wrong way, so we would appreciate it if you could tell us how to use the optuna_scheduler.

**Comment by cvphelps** [comment]
@gtarpenning could you please kindly take a look at the Optuna team's comment here

**Comment by gtarpenning** [comment]
> wandb/sweep-jobs/job-fashion-MNIST-train:latest

Thanks for checking it out! Do you mind trying again, the wandb project has now been made public and should be accessible. Please continue to let me know if there are any other problems! 

You can check out the `wandb` jobs in the wandb.ai project here: https://wandb.ai/wandb/sweep-jobs/jobs

**Pull request merged**